### ISSUE TITLE
Inheritance type propagation problems through `__post_init__`

### ISSUE THREAD

**Issue by janEbert** [astral-sh/ty#2830]
### Summary

`ty` displays an error for the following MRE, even though typing is theoretically correct. It seems like `ty` does not correctly resolve the type through the parent class's method call.

`self.a` is already tested to not be `None` in the parent class's method (`MySuperclass.__post_init__`); however, when an `int`-specific attribute is accessed (`bit_length`), it can be seen that `ty` does not actually propagate/resolve the `is not None`-check from the `super().__post_init__()` call in the child class's method (`MySubclass.__post_init__`).

I tested this using simply `ty check` on an empty project with both 0.0.17 and present `main` (532e761e8f8953b59d6bc61e80ed198473b853c6).

Playground link: https://play.ty.dev/2cbc9fbc-c8c0-4bdd-837c-c1f6dbb07c83

```python
from dataclasses import dataclass


@dataclass
class MySuperclass:
    a: int | None = None

    def __post_init__(self) -> None:
        assert self.a is not None


@dataclass
class MySubclass(MySuperclass):
    def __post_init__(self) -> None:
        super().__post_init__()
        self.a.bit_length()  # <-- error[unresolved-attribute]: Attribute `bit_length` is not defined on `None` in union `int | None`
```

### Version

0.0.17 and 532e761e8f8953b59d6bc61e80ed198473b853c6

**Comment by AlexWaygood** [comment]
Thanks for the report! However, I think ty is behaving correctly here. [Mypy](https://mypy-play.net/?gist=d2b64b0e89c3886648119151fbec42b4), [pyright](https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.14&pythonPlatform=Darwin&reportUnreachable=true&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAoOgAXiTUwxpaygFkBPAZQCuRbCA4YAXDSjSoccQQB25AD5QAcmAXYoAXnWbsdGbGzAoAfXMkMCc-gX5b5gBQ5UwAJRQAtAD59WpLGxqwi5G7AAHRwBHgKYOQaWnQ0jJRi7NR4vIIARmLOOUIiYh5BMjCmFlZgNnYOTq7Y7l5%2BAdjlwVAYxSDOHpGW1rb2jpb9Ul3dzVFwkbljqNgKAOYIABb90gDEUAA83t5QIuAgANoCCiC4YKgAbtgw3ogIIPi5AgjYALryAIIIV7vT7aAAGC1sS1WG1BsSg8XIlWA9keUE0UFBSWwsPsUEu%2BHRoPsKnaoKAA) and [pyrefly](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeS4ATrgLYAEmqALqgMZSpxzx0Q3G5KTBszYcuAHXRSpAAUYt2nOFKVc6AWVIBlAK7EYlNXERS65uqkS90wgD50AcrnQw6AXicuYM9BYYwYHQA%2BsECcEzBEOgQkcEAFNxQYACUdAC0AHxerqZ%2B-hbKhsJJYISovHB06LjCzq6%2BcgpiyqriVVp62MbxnfqGxil5-piBIWG4EVExcYkwyWlZOTDDBeZw-ZTxKYSh4ZHRsaHbZmt0peWE2Eew6ADmTAAW2%2BYAxHQAPOnpdIbUlABtXToSjwXBQABuMEw6WYTEoEGwuiYMAAutYAIJMeGI5FuAAG10itwej3xlWqtQCkFcmDoLjo%2BPqMHJ0TowIgDPx0Xsy3xIAANCBkdA4CRyIgQO8AKpMaCxUh0MDA1hylwqaRYMZgQQ0ZjBdC6GjYQzxfDWHmLbIRSircygpi6Sh%2BMASECOI0m210YD4AC%2BbqkgpAZFBYCgpEITFoUAo7wACqQwxHzhgcAQ6KwXJA7k7mJz0IQpO9tDA3I9scQTAB6auhwIRwiCO7VmDoauYXCsODVrPoHN5tXtpWCSwQ1DQVDYWCZ7MQXOUfMM3DEIdiqRkJ4udJQyhwAseOhugDMhAAjAAmQPoEB%2BoVsOVQgBi0BgFDQWDwRDIt6AA) all issue the same diagnostic.

In general, any type narrowing done inside a method is local to that method and does not propagate outside of that method. We have the same behaviour for `__post_init__` here as we do for `__init__` in this class:

```py
class F:
    def __init__(self, x: int | None):
        self.x = x
        assert self.x is not None
        reveal_type(self.x)  # locally narrowed to `int`

reveal_type(F(42).x)  # revealed: int | None | Unknown
```

Moreover, an explicit type annotation for a variable (`x: int | None` in the dataclass body) always overrides any type we might infer from assignments inside methods:

```py
class F:
    x: int | None

    def __init__(self, x: int):
        self.x = x  # the local inferred type is `int` inside this method

reveal_type(F(42).x)  # revealed: int | None, because of the explicit declaration in the class body
```

The better way of writing your dataclass here would be something like this, I think:

```py
from dataclasses import dataclass, InitVar, field


@dataclass
class MySuperclass:
    _a: InitVar[int | None] = None
    a: int = field(init=False)

    def __post_init__(self, a: int | None) -> None:
        assert a is not None
        self.a = a


@dataclass
class MySubclass(MySuperclass):
    def __post_init__(self, a: int | None) -> None:
        super().__post_init__(a)
        self.a.bit_length()
```

I would advise reading the [dataclass documentation](https://docs.python.org/3/library/dataclasses.html) for more details on how `InitVar` and `field` work.

**Comment by janEbert** [comment]
Thanks a lot for the very detailed response and your kind suggestions. Sorry about taking time from your side when this behavior is much less specific than I assumed; I really should have checked with `mypy` as well before reporting.

**Comment by AlexWaygood** [comment]
No problem at all! Thanks for the clear bug report
